avocado/core/task/statemachine.py

```diff
@@ -74,20 +75,69 @@ async def bootstrap(self):
 
     async def triage(self):
         """Reads from triaging, moves into either: ready or finished."""
+        def fail_on_triage(runtime_task):
+            self._state_machine.finished.append(runtime_task)
+            runtime_task.status = 'FAILED ON TRIAGE'
+
         try:
             async with self._state_machine.lock:
                 runtime_task = self._state_machine.triaging.pop(0)
         except IndexError:
             return
 
-        requirements_ok = await self._spawner.check_task_requirements(runtime_task)
-        if requirements_ok:
+        # check for requirements a task may have
+        requirements_ok = (
+                await self._spawner.check_task_requirements(runtime_task))
+        if not requirements_ok:
             async with self._state_machine.lock:
-                self._state_machine.ready.append(runtime_task)
-        else:
+                fail_on_triage(runtime_task)
+            return
+
+        # handle task dependencies
+        if runtime_task.task.dependencies:
+            # check of all the dependency tasks finished
             async with self._state_machine.lock:
-                self._state_machine.finished.append(runtime_task)
-                runtime_task.status = 'FAILED ON TRIAGE'
+                finished_tasks = [rt_task.task for rt_task
+                                  in self._state_machine.finished]
+            if not runtime_task.task.dependencies.issubset(finished_tasks):
+                async with self._state_machine.lock:
+                    self._state_machine.triaging.append(runtime_task)
+                    runtime_task.status = 'WAITING DEPENDENCIES'
+                await asyncio.sleep(0.1)
+                return
```

Comment:
The fact that the code `return`s here, after means that the entire `triage()` gets executed again later.  This is not wrong, but I think that calling `_spawner.check_task_requirements(runtime_task)` dozens or even hundreds of times is not correct too.

An optimization is to do something like:

```diff
         # handle task dependencies
         if runtime_task.task.dependencies:
-            # check of all the dependency tasks finished
-            async with self._state_machine.lock:
-                finished_tasks = [rt_task.task for rt_task
-                                  in self._state_machine.finished]
-            if not runtime_task.task.dependencies.issubset(finished_tasks):
+            while True:
+                # check of all the dependency tasks finished
                 async with self._state_machine.lock:
-                    self._state_machine.triaging.append(runtime_task)
-                    runtime_task.status = 'WAITING DEPENDENCIES'
+                    finished_tasks = [rt_task.task for rt_task
+                                      in self._state_machine.finished]
+                    if runtime_task.task.dependencies.issubset(finished_tasks):
+                        break
+                runtime_task.status = 'WAITING DEPENDENCIES'
                 await asyncio.sleep(0.1)
-                return
 
             # dependencies finished, let's check if they finished
             # successfully, so we can move on with the parent task
```

Reply:
I see your point, and I agree with it. I was trying to create a flag to let the code know that the spawner part was already okay when a task starts the triage again but left it away.

Unfortunately, your proposal is not working. I'll investigate why and try to develop a solution to avoid the execution of the spawner side check multiple times.